storage/remote/queue_manager.go

```diff
@@ -2136,36 +2164,39 @@ func buildTimeSeries(timeSeries []prompb.TimeSeries, filter func(prompb.TimeSeri
 func compressPayload(tmpbuf *[]byte, inp []byte, enc Compression) (compressed []byte, _ error) {
 	switch enc {
 	case SnappyBlockCompression:
-		compressed = snappy.Encode(*tmpbuf, inp)
 		if n := snappy.MaxEncodedLen(len(inp)); n > len(*tmpbuf) {
 			// grow the buffer for the next time
 			*tmpbuf = make([]byte, n)
 		}
+		compressed = snappy.Encode(*tmpbuf, inp)
 		return compressed, nil
 	default:
 		return compressed, fmt.Errorf("Unknown compression scheme [%v]", enc)
 	}
 }
 
-func buildWriteRequest(logger log.Logger, timeSeries []prompb.TimeSeries, metadata []prompb.MetricMetadata, pBuf *proto.Buffer, buf *[]byte, filter func(prompb.TimeSeries) bool, enc Compression) (compressed []byte, highest, lowest int64, _ error) {
+func buildWriteRequest(logger log.Logger, timeSeries []*prompb.TimeSeries, metadata []*prompb.MetricMetadata, pBuf, buf *[]byte, filter func(*prompb.TimeSeries) bool, enc Compression) (compressed []byte, highest, lowest int64, _ error) {
 	highest, lowest, timeSeries,
 		droppedSamples, droppedExemplars, droppedHistograms := buildTimeSeries(timeSeries, filter)
 
 	if droppedSamples > 0 || droppedExemplars > 0 || droppedHistograms > 0 {
 		level.Debug(logger).Log("msg", "dropped data due to their age", "droppedSamples", droppedSamples, "droppedExemplars", droppedExemplars, "droppedHistograms", droppedHistograms)
 	}
 
-	req := &prompb.WriteRequest{
-		Timeseries: timeSeries,
-		Metadata:   metadata,
+	req := prompb.WriteRequestFromVTPool()
+	req.Timeseries = timeSeries
+	req.Metadata = metadata
+	if pBuf == nil {
+		pBuf = &[]byte{} // For convenience in tests. Not efficient.
 	}
 
-	if pBuf == nil {
-		pBuf = proto.NewBuffer(nil) // For convenience in tests. Not efficient.
-	} else {
-		pBuf.Reset()
+	if len(*pBuf) < req.SizeVT() {
+		*pBuf = make([]byte, req.SizeVT())
 	}
-	err := pBuf.Marshal(req)
+	size, err := req.MarshalToVT(*pBuf)
+	req.Metadata = []*prompb.MetricMetadata{}
+	req.Timeseries = []*prompb.TimeSeries{}
+	req.ReturnToVTPool()
```

Comment:
We need to reset the `req.Metadata` and `req.Timeseries` since the input `timeSeries` and `metadata` will be reused after the function returns. `req.ReturnToVTPool()` will destroy those fields.